### PR TITLE
Remove remaining CRT PAL wrappers and enable including standard headers in the CoreCLR build

### PR DESCRIPTION
**Pull request by jkoritzinsky**
~~Depends on #98238~~

- Remove malloc-family PAL and update callsites that might get passed 0 to bump up to 1.
- Move `getenv` usage to use the `PAL` function directly when on non-Windows (to preserve the existing behavior).
- Remove other remaining CRT PAL shims
- Remove header shims and enable building with the CRT and STL headers across the product, with various build fixes required (mostly around using the standard min/max definitions)

### CODE REVIEW

File: src/coreclr/inc/clr_std/vector
Comment (jkotas) [inline src/coreclr/inc/clr_std/vector:32]: Is switching to STL definition of collection classes going to introduce STL exception handling? The runtime exception handling is not compatible with that.

File: src/coreclr/pal/src/cruntime/misc.cpp
Comment (jkotas) [inline src/coreclr/pal/src/cruntime/misc.cpp:116]: Is it non-trivial to delete this for some reason?

File: src/coreclr/inc/clr_std/vector
Comment (jkoritzinsky) [reply src/coreclr/inc/clr_std/vector:32]: I think it might, but we can't switch to using the CRT headers without switching these headers as well due to some build issues.

I'll keep some of these headers (the ones that are used in the runtime and use components that would use C++ exception handling) and change the namespace to something like `clr_std` to avoid conflicting with the STL.

File: src/coreclr/pal/src/cruntime/misc.cpp
Comment (jkoritzinsky) [reply src/coreclr/pal/src/cruntime/misc.cpp:116]: I missed these when going through the removal. I'll circle back and make sure to remove them.

File: src/coreclr/inc/clr_std/vector
Comment (jkoritzinsky) [reply src/coreclr/inc/clr_std/vector:32]: After investigating further, the only place in the repo that used any of the STL facilities that would introduce exception handling is superpmi. The only usages within the rest of the CoreCLR codebase of the `clr_std` headers were compile-time type traits and `std::move`. Therefore, I believe we can remove all of these headers without issue (and be cognizant in the future around introducing STL usage that may introduce exceptions).

File: src/coreclr/jit/utils.h
Comment (lambdageek) [inline src/coreclr/jit/utils.h:174]: why can't you use `std::min`/`std::max` like elsewhere? Also why rvalue references?

File: src/coreclr/pal/src/include/pal/malloc.hpp
Comment (janvorli) [inline src/coreclr/pal/src/include/pal/malloc.hpp:37]: I wonder if we could fix the places where we call InternalMalloc with 0 size instead and get rid of it.

Comment (jkoritzinsky) [reply src/coreclr/pal/src/include/pal/malloc.hpp:37]: I can look through the PAL and do that. I was trying to limit the PAL-internal changes to keep this a little smaller, but I'll take a look.

File: src/coreclr/vm/methodtablebuilder.cpp
Comment (janvorli) [inline src/coreclr/vm/methodtablebuilder.cpp:8332]: I am not sure why this change and the train of changes resulting from it was needed.

File: src/coreclr/vm/perfmap.cpp
Comment (janvorli) [inline src/coreclr/vm/perfmap.cpp:49]: There is at least one more change like this with conditional path for Windows / Unix. Can we do this inside of the CLRConfigNoCache::Get instead? Or is there a case when we don't want to use the PAL_getenv argument on Unix?

Comment (janvorli) [reply src/coreclr/vm/perfmap.cpp:49]: Ok, makes sense.

File: src/coreclr/inc/allocacheck.h
Comment (janvorli) [inline src/coreclr/inc/allocacheck.h:26]: Why does it have to be excluded for Apple?

Comment (jkoritzinsky) [reply src/coreclr/inc/allocacheck.h:26]: Apple doesn't have the `malloc.h` header. It's Windows and POSIX only. This was working before because we had a malloc.h header in our "CRT shim" header set.

Comment (janvorli) [reply src/coreclr/inc/allocacheck.h:26]: Ah, I didn't know that. So maybe we should unconditionally include stdlib.h instead, that's the correct header for malloc everywhere.

Comment (jkoritzinsky) [reply src/coreclr/inc/allocacheck.h:26]: I'll try out removing this header include and seeing which places need it for the POSIX apis it defines.

File: src/coreclr/debug/inc/dbgipcevents.h
Comment (jkotas) [inline src/coreclr/debug/inc/dbgipcevents.h:771]: It is more common to use `(TADDR)0` or just `0` for these situations.

File: src/coreclr/inc/allocacheck.h
Comment (jkoritzinsky) [reply src/coreclr/inc/allocacheck.h]: Looks like it was only needed on Windows for `_alloca`.

File: src/coreclr/vm/methodtablebuilder.cpp
Comment (jkoritzinsky) [reply src/coreclr/vm/methodtablebuilder.cpp]: Removed this set of changes

**Comment by jkoritzinsky** [comment]
Failures are unrelated. This is ready for another round of review!

**Comment by janvorli** [comment]
I will review it tomorrow. Also please run the outer loop tests to verify that PAL tests are still passing.

File: src/coreclr/gc/vxsort/defs.h
Comment (jkotas) [inline src/coreclr/gc/vxsort/defs.h:48]: Delete the custom definition of max/min to avoid this undef?

File: src/coreclr/inc/daccess.h
Comment (jkotas) [inline src/coreclr/inc/daccess.h:564]:
```suggestion
#ifndef HOST_WINDOWS
```
This is under `#ifndef NATIVEAOT`

File: src/coreclr/inc/daccess.h
Comment (jkotas) [inline src/coreclr/inc/daccess.h:576]: What is `PAL_STDCPP_COMPAT` for after this change? Is it still necessary?

File: src/coreclr/pal/inc/pal_mstypes.h
Comment (jkotas) [inline src/coreclr/pal/inc/pal_mstypes.h:68]:
```suggestion
```

File: src/coreclr/CMakeLists.txt
Comment (janvorli) [inline src/coreclr/CMakeLists.txt:222]: Why do we replace all the compile definitions by "RC_INVOKED" here?

File: src/coreclr/clrdefinitions.cmake
Comment (janvorli) [inline src/coreclr/clrdefinitions.cmake:56]: Why do we use `add_compile_definitions` here and `add_definitions` above?

File: src/coreclr/CMakeLists.txt
Comment (janvorli) [inline src/coreclr/CMakeLists.txt:208]: Why do we have AppleClang as a separate compiler id?

File: src/coreclr/clrdefinitions.cmake
Comment (jkoritzinsky) [reply src/coreclr/clrdefinitions.cmake:56]: I generally want to move us to `add_compile_definitions` in general as it's a little clearer (don't need the `-D` parameter) and propagates well to our custom functions logic like `get_compile_definitions`. I didn't want to bloat this PR further by changing all of the other ones in this PR.

Comment (janvorli) [reply src/coreclr/clrdefinitions.cmake:56]: Ok, it makes sense then.

File: src/coreclr/CMakeLists.txt
Comment (jkoritzinsky) [reply src/coreclr/CMakeLists.txt:208]: We need to specify it separately here as CMake's generator expressions don't allow us to do a regex match for the compiler ID; CMake only does exact matches here.

File: eng/native/functions.cmake
Comment (janvorli) [inline eng/native/functions.cmake:223]: I wonder if this was a compensation for the COMPILE_DEFINITIONS property that you have fixed based on my feedback. Or is this something else?

File: src/coreclr/CMakeLists.txt
Comment (janvorli) [reply src/coreclr/CMakeLists.txt:208]: I have actually meant something else - why do we have AppleClang here at all - this seems to be the only usage of AppleClang in the whole build system.

File: eng/native/functions.cmake
Comment (jkoritzinsky) [reply eng/native/functions.cmake:223]: The `get_source_file_property` was added to enable getting source-file-specific definitions. In particular, RC_INVOKED. The code your feedback was on was the pair to this line.

File: src/coreclr/CMakeLists.txt
Comment (jkoritzinsky) [reply src/coreclr/CMakeLists.txt:208]: In the rest of the build system, we use `CMAKE_<LANG>_COMPILER_ID MATCHES "Clang"` to match both Clang and AppleClang. As of CMake 3.0 (and CMake policy [CMP0025](https://cmake.org/cmake/help/latest/policy/CMP0025.html)), Apple's Clang build is a different ID. When we updated to a minimum CMake >= 3.0, we were able to use the `MATCHES` model for all cases where we were checking for Clang to easily support AppleClang.

File: src/coreclr/gc/gc.cpp
Comment (AaronRobinsonMSFT) [inline src/coreclr/gc/gc.cpp:43397]: This looks wrong. We've lost a `*1024`.

File: src/coreclr/gc/gc.cpp
Comment (AaronRobinsonMSFT) [inline src/coreclr/gc/gc.cpp:51273]: The `u` should no longer be needed if we are doing a cast.

Several places in this file do that. I would do a cast or use the `u`, but not both.

File: src/coreclr/gc/gcpriv.h
Comment (AaronRobinsonMSFT) [inline src/coreclr/gc/gcpriv.h:5975]: There is a lot of subtlty going on with this code. What is the cost of not doing this? I am concerned this is going to introduce subtle bugs.

File: src/coreclr/hosts/coreshim/CoreShim.h
Comment (AaronRobinsonMSFT) [inline src/coreclr/hosts/coreshim/CoreShim.h:8]: Is this now being inherited?

File: src/coreclr/jit/hashbv.cpp
Comment (AaronRobinsonMSFT) [inline src/coreclr/jit/hashbv.cpp:827]:
```suggestion
        indexType   bits_to_set = min((unsigned)BITS_PER_NODE, numToSet - i);
```

Most of this review uses `(unsigned)`. I prefer `unsigned int`, but I am fine with either as long as we are consistent.

File: src/coreclr/jit/utils.h
Comment (AaronRobinsonMSFT) [inline src/coreclr/jit/utils.h:173]: Another instance of this pattern. I do not think we should be introducing yet another min/max pattern that is unexpected. What is the cost to properly cast?

File: src/coreclr/nativeaot/Runtime/threadstore.cpp
Comment (AaronRobinsonMSFT) [inline src/coreclr/nativeaot/Runtime/threadstore.cpp:233]:
```suggestion
    int l = iteration >= 0 ? min(iteration, 30) : 30;
```
The converting of `-1` to a max unsigned is a lot of cleverness for one line.

File: src/coreclr/pal/tests/palsuite/runpaltests.sh
Comment (AaronRobinsonMSFT) [inline src/coreclr/pal/tests/palsuite/runpaltests.sh:12]: Is this change correct?

File: src/coreclr/tools/superpmi/superpmi-shared/standardpch.h
Comment (AaronRobinsonMSFT) [inline src/coreclr/tools/superpmi/superpmi-shared/standardpch.h:76]: What is the cost to remove this pattern?

File: src/coreclr/vm/eetwain.cpp
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/eetwain.cpp:433]: We are mixing `unsigned int` and `unsigned`. If I recall, `unsigned` implies `int`, why the inconsistency?

Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/eetwain.cpp:433]: Same below.

**Comment by AaronRobinsonMSFT** [comment]
Are we going to defer `USE_STL` usage until after this work is in?

File: src/coreclr/hosts/coreshim/CoreShim.h
Comment (jkoritzinsky) [reply src/coreclr/hosts/coreshim/CoreShim.h:8]: Yes this is now specified at the command line so we don't need to do it in source.

File: src/coreclr/gc/gcpriv.h
Comment (jkoritzinsky) [reply src/coreclr/gc/gcpriv.h:5975]: These min/max overloads allow us to use (same-signedness) integer promotion rules in our min/max usage, which is the vast majority of the usage in the GC and the JIT (the other place I have a copy of these). Differing-signedness promotion is detected and produces an error in our GCC build leg (and the Win-x86 build leg also for some reason). These locations relied on the macro-based prior implementation to allow all of the integer promotion rules.

I could move every single min/max call to have exact type matches with manual casts for temporaries, but the number of places would quickly become a large portion of the changed lines in this PR and would make the code more difficult to read.

File: src/coreclr/jit/hashbv.cpp
Comment (jkoritzinsky) [reply src/coreclr/jit/hashbv.cpp:827]: I was trying to match the surrounding types. In most of the other cases, `unsigned` was used for one of the input types or the result's type. This is one of the few cases where one of the inputs was `unsigned int` instead of `unsigned`. I also prefer `unsigned int`, but I was deferring to the existing style in each function.

File: src/coreclr/gc/gcpriv.h
Comment (jkoritzinsky) [reply src/coreclr/gc/gcpriv.h:5975]: I'm taking a look at removing these now. It's not as bad now that I've already had to adjust for signedness, so it's more just tedious. It's doable though, so I'm going through it.

**Review by AaronRobinsonMSFT — APPROVED** [review]
Two small nits. Other than that, this LGTM.

**Comment by janvorli** [comment]
I'm taking a last look now.

**Review by janvorli — APPROVED** [review]
LGTM modulo the comment.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by jkoritzinsky** [comment]
/ba-g Timeouts look like the slow mac issue, all with non-CoreCLR runtimes.

**Comment by BruceForstall** [comment]
Amazing. I never thought this would happen.

Could (should) someone convert the JIT from the jitstd implementations of `list` and `vector` (and others) to the platform implementations?

Is there guidance about what parts of the STL can and can't be used?

**Comment by AaronRobinsonMSFT** [comment]
> Is there guidance about what parts of the STL can and can't be used?

Not at present, but we need to avoid cases where exceptions are thrown. This means no STL collections should be used in the product. Using in tests is fine, but we need to make sure the code isn't shared in some non-obvious manner.

If someone wanted to use the STL collections the first step would be to write a custom allocator so we can control the failure mode. Ideally we would typedef all the STL collections to be something like `clr_vector` that uses the custom allocator. There are also other APIs like `std::make_shared<>`, which can have unexpected behavior. `std::string` is also a type of "collection" and falls under the same bucket.

My preference, no way to enforce this, would be for me and a group of others to be added to PRs where new uses of the STL are used. After a while we will all agree on approved patterns and then we can document them appropriately.

**Comment by jkoritzinsky** [comment]
I'll work on a doc for guidance for STL usage with some useful typedefs.

**Pull request merged**
